pageserver/src/tenant/timeline.rs

```diff
@@ -631,6 +633,66 @@ impl Timeline {
             return Ok(());
         }
 
+        // retry two times to allow first round to find layers which need to be downloaded, then
+        // download them, then retry compaction
+        for round in 0..ROUNDS {
+            // should we error out with the most specific error?
+            let last_round = round == ROUNDS - 1;
+
+            let res = self.compact_without_ondemand_downloads(ctx).await;
```

Comment:
Should a subcontext be created which would ERROR on ... downloads.

No, and but the name needs to be changed, because it can on-demand download for image creation, shown by the added test :)

Reply:
Renamed in eb95d489 as `compact_inner`.